### PR TITLE
Standard provider python operator

### PR DESCRIPTION
**Pull request by gopidesupavan**
Moving python operators/sensors to standard provider

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: tests/system/providers/apache/kafka/example_dag_event_listener.py [hunk 1]
```diff
@@ -26,12 +26,12 @@
...
 from airflow.models import Connection
+from airflow.providers.apache.kafka.operators.produce import ProduceToTopicOperator
+from airflow.providers.apache.kafka.sensors.kafka import AwaitMessageTriggerFunctionSensor
 
 # This is just for setting up connections in the demo - you should use standard
 # methods for setting these connections in production
-from airflow.operators.python import PythonOperator
-from airflow.providers.apache.kafka.operators.produce import ProduceToTopicOperator
-from airflow.providers.apache.kafka.sensors.kafka import AwaitMessageTriggerFunctionSensor
+from airflow.providers.standard.core.operators.python import PythonOperator
```
Comment (gopidesupavan) [inline tests/system/providers/apache/kafka/example_dag_event_listener.py]: Python and Bash operators are currently used in several system tests as well as tests for other providers, do we ignore doing refactoring system test imports until standard provider releases, or any other approach to work these imports?

**Comment by gopidesupavan** [comment]
cc: @romsharon98

File: tests/system/providers/apache/kafka/example_dag_event_listener.py (hunk 1 above)
Comment (gopidesupavan) [reply tests/system/providers/apache/kafka/example_dag_event_listener.py]: or should i add something like this 
```
try:
    from airflow.providers.standard.core.operators.python import PythonOperator
except ImportError:
    from airflow.operators.python import PythonOperator
```

**Comment by gopidesupavan** [comment]
Some of the tests are failing due to compatibility issues. It appears that the changes related to the `triggering_asset_events `event are associated with Airflow 3.0.0. Since we're transitioning the Python operator into the providers' ecosystem, the standard provider now requires a minimum Airflow version of 2.10.0. As a result, the compatibility tests for version 2.10.1 are failing. Any suggestions ? 


 https://github.com/apache/airflow/actions/runs/11308098417/job/31450685352?pr=42081#step:11:12372

**Comment by gopidesupavan** [comment]
@potiuk It appears that this PR needs to be initiated from the Apache repository, or it should be someone with committer privileges. The PR involves modifications in the `dev/breeze` and `dev/tests_common/test_utils/compat.py` sections, but it seems that the CI process is overriding these scripts with those from the main branch, which could be causing some issues. 

Here it is removing https://github.com/apache/airflow/actions/runs/11308098408/job/31450591466#step:4:219

https://github.com/apache/airflow/pull/42081/files#diff-cbadf25516149c7fe3d19e45b5b03789f4bb93bd6d837eaa365dfd0ffb771a0fR55

For `dev/tests_common/test_utils/compat.py` have workaround but dont like this, i have imported operators with try/catch like this https://github.com/apache/airflow/pull/42081/files#diff-a842cdd5bcf15ce8859d6115b48853c3ac654701a3279bce73969be8922dbbe8R47. these imports can be replaced using compat.

Any thoughts please ?

**Comment by gopidesupavan** [comment]
> @potiuk It appears that this PR needs to be initiated from the Apache repository, or it should be someone with committer privileges. The PR involves modifications in the `dev/breeze` and `dev/tests_common/test_utils/compat.py` sections, but it seems that the CI process is overriding these scripts with those from the main branch, which could be causing some issues.
> 
> Here it is removing https://github.com/apache/airflow/actions/runs/11308098408/job/31450591466#step:4:219
> 
> https://github.com/apache/airflow/pull/42081/files#diff-cbadf25516149c7fe3d19e45b5b03789f4bb93bd6d837eaa365dfd0ffb771a0fR55
> 
> For `dev/tests_common/test_utils/compat.py` have workaround but dont like this, i have imported operators with try/catch like this https://github.com/apache/airflow/pull/42081/files#diff-a842cdd5bcf15ce8859d6115b48853c3ac654701a3279bce73969be8922dbbe8R47. these imports can be replaced using compat.
> 
> Any thoughts please ?

looks like the tests are going through fine now.

> @potiuk It appears that this PR needs to be initiated from the Apache repository, or it should be someone with committer privileges. The PR involves modifications in the `dev/breeze` and `dev/tests_common/test_utils/compat.py` sections, but it seems that the CI process is overriding these scripts with those from the main branch, which could be causing some issues.
> 
> Here it is removing https://github.com/apache/airflow/actions/runs/11308098408/job/31450591466#step:4:219
> 
> https://github.com/apache/airflow/pull/42081/files#diff-cbadf25516149c7fe3d19e45b5b03789f4bb93bd6d837eaa365dfd0ffb771a0fR55
> 
> For `dev/tests_common/test_utils/compat.py` have workaround but dont like this, i have imported operators with try/catch like this https://github.com/apache/airflow/pull/42081/files#diff-a842cdd5bcf15ce8859d6115b48853c3ac654701a3279bce73969be8922dbbe8R47. these imports can be replaced using compat.
> 
> Any thoughts please ?

looks like tests are running..now..

File: tests/system/providers/apache/kafka/example_dag_event_listener.py (hunk 1 above)
Comment (gopidesupavan) [reply tests/system/providers/apache/kafka/example_dag_event_listener.py]: This is not required imports. it is working without try catch.

**Comment by gopidesupavan** [comment]
Compat tests 2.10.1 failing because python operator has changes related to airflow 3.0.0, should i update back changes to support in 2.10.X?

**Comment by potiuk** [comment]
> Compat tests 2.10.1 failing because python operator has changes related to airflow 3.0.0, should i update back changes to support in 2.10.X?

It looks like it's more of "unit test compatibility" than backporting it. I think in this case simply tests should be fixed to be able to handle past providers. We do not need to backport this change, but we need to make sure that the assertions work when `airflow` is installed from "2.10.1"

**Comment by gopidesupavan** [comment]
> > Compat tests 2.10.1 failing because python operator has changes related to airflow 3.0.0, should i update back changes to support in 2.10.X?
> 
> It looks like it's more of "unit test compatibility" than backporting it. I think in this case simply tests should be fixed to be able to handle past providers. We do not need to backport this change, but we need to make sure that the assertions work when `airflow` is installed from "2.10.1"

thanks jarek, When installing Airflow 2.10.1, the Python operator is included as part of the core package. Should assertions be checked against the core Python operator (i.e., airflow.operators.python) or against airflow.providers.standard.operators.python? Opting for the former would require code changes due to the adjustments already made for Airflow 3. Do you agree that if Airflow 2.10.1 is installed, validation tests should be conducted on airflow.operators.python, as long as the tests remain compatible?

**Comment by gopidesupavan** [comment]
> > > Compat tests 2.10.1 failing because python operator has changes related to airflow 3.0.0, should i update back changes to support in 2.10.X?
> > 
> > 
> > It looks like it's more of "unit test compatibility" than backporting it. I think in this case simply tests should be fixed to be able to handle past providers. We do not need to backport this change, but we need to make sure that the assertions work when `airflow` is installed from "2.10.1"
> 
> thanks jarek, When installing Airflow 2.10.1, the Python operator is included as part of the core package. Should assertions be checked against the core Python operator (i.e., airflow.operators.python) or against airflow.providers.standard.operators.python? Opting for the former would require code changes due to the adjustments already made for Airflow 3. Do you agree that if Airflow 2.10.1 is installed, validation tests should be conducted on airflow.operators.python, as long as the tests remain compatible?

ah i got it what you mean by 'assertions work', need to make some adjustments in the python operator tests related to new feature updates. hope that works..

**Comment by potiuk** [comment]
> ah i got it what you mean by 'assertions work', need to make some adjustments in the python operator tests related to new feature updates. hope that works..

Yes. Some examples and detailed explanation on how to do this "test_compatibility" are here: https://github.com/apache/airflow/blob/main/contributing-docs/testing/unit_tests.rst#implementing-compatibility-for-provider-tests-for-older-airflow-versions

File: providers/src/airflow/providers/amazon/aws/operators/appflow.py [hunk 2]
```diff
@@ -21,12 +21,16 @@
...
-from airflow.operators.python import ShortCircuitOperator
 from airflow.providers.amazon.aws.hooks.appflow import AppflowHook
 from airflow.providers.amazon.aws.operators.base_aws import AwsBaseOperator
 from airflow.providers.amazon.aws.utils import datetime_to_epoch_ms
 from airflow.providers.amazon.aws.utils.mixins import AwsBaseHookMixin, AwsHookParams, aws_template_fields
 
+try:
+    from airflow.providers.standard.operators.python import ShortCircuitOperator
+except ImportError:
+    from airflow.operators.python import ShortCircuitOperator  # type: ignore[no-redef,attr-defined]
```
Comment (mobuchowski) [inline providers/src/airflow/providers/amazon/aws/operators/appflow.py]: Maybe add those imports to `compat` provider and then you wouldn't have to try/except everywhere? 

The drawback is you'd need to add dependency to `compat` providers.

Comment (gopidesupavan) [reply providers/src/airflow/providers/amazon/aws/operators/appflow.py]: Sure i will check, if that is possible i will add.

**Comment by gopidesupavan** [comment]
I believe we have to move these scripts as well into standard provider `airflow/utils/python_virtualenv.py`  `airflow/utils/python_virtualenv_script.jinja2` ?

**Comment by potiuk** [comment]
> I believe we have to move these scripts as well into standard provider `airflow/utils/python_virtualenv.py` `airflow/utils/python_virtualenv_script.jinja2` ?

Yes. Absolutely.

**Comment by gopidesupavan** [comment]
> > I believe we have to move these scripts as well into standard provider `airflow/utils/python_virtualenv.py` `airflow/utils/python_virtualenv_script.jinja2` ?
> 
> Yes. Absolutely.

cool, Thank you :)

File: providers/src/airflow/providers/amazon/aws/operators/appflow.py (hunk 2 above)
Comment (potiuk) [reply providers/src/airflow/providers/amazon/aws/operators/appflow.py]: Yeah. Good idea.  Just make sure to add "compat>=NEXT_MINOR_VERSION" for the compat dependency to "standard"

**Review by potiuk — APPROVED** [review]
Just one thing about compat provider reuse.

**Comment by gopidesupavan** [comment]
> Just one thing about compat provider reuse.

yeah am updating imports to use compat.

File: providers/src/airflow/providers/amazon/aws/operators/appflow.py (hunk 2 above)
Comment (eladkal) [reply providers/src/airflow/providers/amazon/aws/operators/appflow.py]: This import is temporary till we first release the standard provider.
I am fine with just adding a comment to drop this try/catch block when provider min version is Airflow 3 (I assume that standard provider will be pre-installed).

Comment (potiuk) [reply providers/src/airflow/providers/amazon/aws/operators/appflow.py]: Actually, we will be able to drop that at the moment we release standard provider.  We do not need to wait for Airflow 3 to happen. 

The idea is that standard provider will be installable - as any other provider in Airflow 2 (with same min-airflow-versions all the other providers). And it does not even have to be pre-installed in Airflow 2, it's enough that the providers that import standard-provider's imports - add the standard provider as a dependency. That's enough.

Comment (eladkal) [reply providers/src/airflow/providers/amazon/aws/operators/appflow.py]: That depends if aws provider will have min version of Airflow 3 or not (or if standard provider will be compatible with older versions)

Comment (potiuk) [reply providers/src/airflow/providers/amazon/aws/operators/appflow.py]: > That depends if aws provider will have min version of Airflow 3 or not (or if standard provider will be compatible with older versions)

No - standard provider is supposed to be compatible with Airlfow 2 and you won't need min-airflow-version 3 to use it. It's entirely possible (and desired) for standard provider to have min-airlfow-version 2.8 or 2.9 or 2.10 (depending when we release it), and it is completely no problem for future AWS provider to import from it even when min-airflow-version for it will be 2.8, 2,9 or 2.10 respectively.

This is all entirely doable and standard provider release was going to be done way before Airflow 3 to suppport that scenario.

**Comment by potiuk** [comment]
> > > Just one thing about compat provider reuse.
> > 
> > 
> > yeah am updating imports to use compat.
> 
> I've added the imports to the compat provider, but I've noticed test failures occurring in selective checks. This is likely because all provider imports are now referenced directly from the compat provider. So, when we revert to standard provider direct imports, we may need to update all the selective check tests accordingly.
> 

No - it's ok - you need to fix the tests accordingly. Those tests are testing whether the right set of "dependent" providers are selected when selective checks is running - so when you start depending on the standard provider and when standard provider depends on some other provider., the tests start to expect changes in the depednencies, so what needs to be done is to update the tests to reflect that.

**Comment by potiuk** [comment]
And it's not that difficult, most of the code is common in constants, and some other could be replaced with search/replace.

**Comment by gopidesupavan** [comment]
> > > > Just one thing about compat provider reuse.
> > > 
> > > 
> > > yeah am updating imports to use compat.
> > 
> > 
> > I've added the imports to the compat provider, but I've noticed test failures occurring in selective checks. This is likely because all provider imports are now referenced directly from the compat provider. So, when we revert to standard provider direct imports, we may need to update all the selective check tests accordingly.
> 
> No - it's ok - you need to fix the tests accordingly. Those tests are testing whether the right set of "dependent" providers are selected when selective checks is running - so when you start depending on the standard provider and when standard provider depends on some other provider., the tests start to expect changes in the depednencies, so what needs to be done is to update the tests to reflect that.

Sure , thanks jarek. Will push the changes..

**Comment by gopidesupavan** [comment]
> BTW. ... some conflicts to solve - so you need to rebase too

Thanks jarek, finally this is green now.

**Review by potiuk — APPROVED** [review]
LGTM. I'd love few other eyes to take a look at this necessarily big change, but it looks cool!

File: providers/src/airflow/providers/standard/operators/python.py [hunk 3]
```diff
@@ -417,14 +393,15 @@ class _BasePythonVirtualenvOperator(PythonOperator, metaclass=ABCMeta):
...
+
     AIRFLOW_SERIALIZABLE_CONTEXT_KEYS = {
         "macros",
         "conf",
         "dag",
         "dag_run",
         "task",
         "params",
-        "triggering_asset_events",
+        "triggering_asset_events" if AIRFLOW_V_3_0_PLUS else "triggering_dataset_events",
```
Comment (o-nikolas) [inline providers/src/airflow/providers/standard/operators/python.py:401]: Why are there dataset events changes in this PR about moving the Python Operator?

File: providers/tests/standard/operators/test_python.py [hunk 4]
```diff
@@ -1554,6 +1556,9 @@ def f():
 class TestExternalPythonOperator(BaseTestPythonVirtualenvOperator):
     opcls = ExternalPythonOperator
 
+    def setup_method(self):
+        logging.config.dictConfig(DEFAULT_LOGGING_CONFIG)
```
Comment (o-nikolas) [inline providers/tests/standard/operators/test_python.py:1561]: Why is this required now?

**Review by o-nikolas — APPROVED** [review]
It's a long PR that's very mechanical to review, but it looks reasonable. I only left a couple clarifying questions.

File: providers/src/airflow/providers/standard/operators/python.py (hunk 3 above)
Comment (gopidesupavan) [reply providers/src/airflow/providers/standard/operators/python.py:401]: The main branch already includes asset event changes compatible with Airflow 3. Since the standard provider needs to support Airflow 2.10, which includes the dataset event feature, I’ve updated this to ensure compatibility with both versions.

File: providers/tests/standard/operators/test_python.py (hunk 4 above)
Comment (gopidesupavan) [reply providers/tests/standard/operators/test_python.py:1561]: There are tests which asserts log messages in `caplog.messages`, some how it worked earlier when these operators are in core area, after moving to providers the caplog.messages asserts are failing, so setting up that worked. 

build logs gone unable to retrieve to share.

Comment (potiuk) [reply providers/tests/standard/operators/test_python.py:1561]: Caplog is evil :)

File: providers/src/airflow/providers/standard/operators/python.py (hunk 3 above)
Comment (potiuk) [reply providers/src/airflow/providers/standard/operators/python.py:401]: Yeah. When Python operator wa spart of "Airflow" it was never supposed to be run against old airflow version (it was impossible since it was part of "airflow"), but when we move to provider, we also run it against previous airflow versions :)

File: providers/src/airflow/providers/standard/operators/python.py [hunk 5]
```diff
@@ -732,12 +709,10 @@ def __init__(
                 f"Sys version: {sys.version_info}. Virtual environment version: {python_version}"
             )
         if python_version is not None and not isinstance(python_version, str):
-            warnings.warn(
-                "Passing non-string types (e.g. int or float) as python_version "
-                "is deprecated. Please use string value instead.",
-                RemovedInAirflow3Warning,
-                stacklevel=2,
+            raise AirflowException(
+                "Passing non-string types (e.g. int or float) as python_version not supported"
```
Comment (jscheffl) [inline providers/src/airflow/providers/standard/operators/python.py:710]: As standard provider is installed as transitive dependency of other providers also in Airflow 2.x... and these are using the new operators... I'd propose we keep the deprecation warning for now. Or we need to conditionally raise if we are running in Airflow 3+ and just emit a warning in Airflow 2?

**Review by jscheffl — APPROVED** [review]
In general great PR, just have one doubt regarding deprecation warning. But not blocking, just double-checking.

@potiuk WDYT... do we need to keep the warning when running in Airflow 2?

File: providers/src/airflow/providers/standard/operators/python.py (hunk 5 above)
Comment (potiuk) [reply providers/src/airflow/providers/standard/operators/python.py:710]: It's ok to drop it I think. For Airflow 2 people will ue "core embedded" Python operator if they don't change import. So no impact for them. If they change Import, they **should** expect changes and changing python version to string will be no problem then.

The whole idea is that once they move the import  in Airflow 2 they **should** get the "full" Airflow 3 experience - and this way they can fix not only deprecations, but also errors they might have when they move. 

The only other operator  that uses PythonVirtualenvOperator as base is BranchPythonVirtualenvOperator - that is also moved here, so I see no real problem with removing deprecation and replacing it with error.

**Review by romsharon98 — APPROVED** [review]

File: providers/src/airflow/providers/standard/operators/python.py (hunk 5 above)
Comment (gopidesupavan) [reply providers/src/airflow/providers/standard/operators/python.py:710]: I remember now—there was a discussion about making imports function properly when a user tries to import airflow.operators.python, redirecting it to airflow.providers.standard.operators.python using something like a meta path, when airflow 3.0 released.

**Comment by gopidesupavan** [comment]
One failure related to spell check this will fix https://github.com/apache/airflow/pull/43538

**Comment by gopidesupavan** [comment]
Are we good to merge this ? 😄

**Comment by potiuk** [comment]
Go ahead and do the honors @gopidesupavan -> That's an interesting one as first(?) serious one to merge as a committer.. Don't be shy. We can always revert it if it breaks things ;)

**Comment by gopidesupavan** [comment]
> Go ahead and do the honors @gopidesupavan -> That's an interesting one as first(?) serious one to merge as a committer.. Don't be shy. We can always revert it if it breaks things ;)

woohoooo merged thank you all for the help and review here, really appreciate it 😄
